test/jdk/jdk/internal/util/ArraysSupport/NewLength.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8247373
+ * @modules java.base/jdk.internal.util
+ * @run testng NewLength
+ * @summary Test edge cases of ArraysSupport.newLength
+ */
+
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+import jdk.internal.util.ArraysSupport;
+
+public class NewLength {
+    static final int IMAX = Integer.MAX_VALUE;
+    static final int SOFT = ArraysSupport.SOFT_MAX_ARRAY_LENGTH;
+
+    // Data that is expected to return a valid value.
+
+    @DataProvider(name = "valid")
+    public Object[][] validProvider() {
+        return new Object[][] {
+           // old     min     pref    expected
+            { 0,      1,      0,      1      },
+            { 0,      1,      2,      2      },
+            { 0,      2,      1,      2      },
+            { 0,      1,      SOFT-1, SOFT-1 },
+            { 0,      1,      SOFT,   SOFT   },
+            { 0,      1,      SOFT+1, SOFT   },
+            { 0,      1,      IMAX,   SOFT   },
+            { 0,      SOFT-1, IMAX,   SOFT   },
+            { 0,      SOFT,   IMAX,   SOFT   },
+            { 0,      SOFT+1, IMAX,   SOFT+1 },
+            { SOFT-2, 1,      2,      SOFT   },
+            { SOFT-1, 1,      2,      SOFT   },
+            { SOFT,   1,      2,      SOFT+1 },
+            { SOFT+1, 1,      2,      SOFT+2 },
+            { IMAX-2, 1,      2,      IMAX-1 },
+            { IMAX-1, 1,      2,      IMAX   },
+            { SOFT-2, 1,      IMAX,   SOFT   },
+            { SOFT-1, 1,      IMAX,   SOFT   },
+            { SOFT,   1,      IMAX,   SOFT+1 },
+            { SOFT+1, 1,      IMAX,   SOFT+2 },
+            { IMAX-2, 1,      IMAX,   IMAX-1 },
+            { IMAX-1, 1,      IMAX,   IMAX   }
+        };
+    }
+
+    // Data that should provoke an OutOfMemoryError
+
+    @DataProvider(name = "error")
+    public Object[][] errorProvider() {
+        return new Object[][] {
+            // old    min   pref
+            {    1,   IMAX, IMAX },
+            { SOFT,   IMAX, 0    },
+            { SOFT,   IMAX, IMAX },
+            { IMAX-1,    2, 0    },
+            { IMAX,      1, 0    },
+            { IMAX,   IMAX, 0    },
+            { IMAX,   IMAX, IMAX }
+        };
+    }
+
+    @Test(dataProvider = "valid")
+    public void valid(int old, int min, int pref, int expected) {
+        assertEquals(ArraysSupport.newLength(old, min, pref), expected);
+    }
+
+    @Test(dataProvider = "error")
+    public void error(int old, int min, int pref) {
+        try {
+            int r = ArraysSupport.newLength(old, min, pref);
+            fail("expected OutOfMemoryError, got normal return value of " + r);
+        } catch (OutOfMemoryError oome) {
+            // ok
```

Comment:
Instead of an //ok comment, I like to give the exception a name that makes the intent clear.

} catch (IllegalArgumentException success) {}